File: LeetCode_Old/CPP/Construct Binary Tree from Inorder and Postorder Traversal.cpp
//
//Created on April 21, 2014
//
//@author: Xufan Du
//

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode *buildTree(vector<int> &inorder, vector<int> &postorder) {
        if(postorder.empty() || inorder.empty()) return NULL;
        return b(inorder,0,inorder.size()-1,postorder,0,postorder.size()-1);
    }
    
    TreeNode *b(vector<int> &inorder, int si,int ei, vector<int> &postorder, int sp,int ep){
        if(sp>ep || si>ei) return NULL;
        TreeNode *root = new TreeNode(postorder[ep]);
        if(sp==ep) return root;
        int in=0;
        for(int i=si;i<=ei;i++){
            if(inorder[i]==postorder[ep]){
                in=i;
                break;
            }
        }
        
        root->left = b(inorder,si,in-1,postorder,sp,sp+in-1-si);
        root->right = b(inorder,in+1,ei,postorder,sp+in-si,ep-1);
        return root;
    }
};

File: LeetCode/106. Construct Binary Tree from Inorder and Postorder Traversal/construct_binary_tree_from_inorder_and_postorder_traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:            
        if not inorder or not postorder:
            return None
        root_val = postorder[-1]
        loc = inorder.index(root_val)
        root = TreeNode(root_val)
        root.left = self.buildTree(inorder[:loc], postorder[:loc])
        root.right = self.buildTree(inorder[loc + 1:], postorder[loc:-1])
        return root
